accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+signal handler.
+
+### Step into / out of
+
+Stepping into a function means pausing execution on the first instruction
+_inside_ a function that is about to be called. We can accomplish this by
+setting a temporary breakpoint on that first instruction, but first we need to
+figure out the address of the function we are about to step into. For direct
+function calls, we can set a breakpoint at the start of the static callee. For
+indirect calls, as long as we pause immediately before the indirect branch, we
+can see which func index we are about to call and set a breakpoint accordingly.
+
+Alternatively, as we are already inserting various bookkeeping function calls
+elsewhere in the function body during code generation, we could track one bit of
+global information that indicates whether or not a function should break
+immediately on entry, checking and potentially clearing that bit at every
+function entry. This would allow us to step into functions by setting that bit
+while in single-stepping mode before processing a call instruction, and then
+clearing that bit once the next function has been entered. This implementation
+adds some overhead by requiring that we are already stepping before the call
+instruction is processed, but does simplify implementation details by not
+requiring special handling of indirect calls.
+
+Stepping out of a function is a bit easier to implement, as it will require us
+to stop execution at any instruction that could cause the function to return
+(branches with a sufficiently large stack depth, or a return instruction), and
+then single-step out. For tail calls we have multiple options: we could treat
+stepping out as breaking when the next function is entered, or we could break
+once we finally return from the tail callee.
+
+### Winch Considerations
+
+Interactive debugging support in an optimizing compiler like Cranelift would
+require fundamental changes to how optimizations are applied and how program
+state is tracked throughout compilation. Maintaining compiler performance while
+integrating those changes would be a significant undertaking. For this reason,
+for the foreseeable future, we are only going to support live debugging with
+Wasmtime’s baseline compiler: Winch.
+
+Winch will require some additional features to enable the debugging workflows
+outlined above:
+
+* The visitor that Winch uses to translate Wasm opcodes to native code will need
+  to be augmented to enable the interleaving of calls to local utility functions
+  that check for breakpoint/watchpoint status.
+
+* We will need to either maintain a mapping from locals to register/stack slot
+  for all Wasm PC points that we might hit a breakpoint or watchpoint, or force
+  winch to unconditionally spill locals to the stack. The latter would greatly
+  simplify tracking local locations, while the former would greatly increase the
+  amount of context we would need to pass into the utility function for
+  inspecting the execution state.
```

Comment:
If I get a vote, I'd say spill to the stack every time, but still keep the current behavior for when debugging is not enabled.

At a later date, we can implement some sort of deoptimizer that transform every function into a trampoline function that either calls the non-debugging-instrumented code, or lazily generates a debugging-instrumented code and calls that instead.  This way we can reduce the overhead for everything that we're not interested in debugging (e.g. if we're running code normally without hitting any breakpoint, the `debug_trap()` function should not be emitted; as soon as a breakpoint is set, the deoptimized/debuggable version of that function is emitted and the `debug_trap()` gets to be called, and functions will be re-generated with debugging instrumentation lazily as we reach them.)

Reply:
Yes, the plan is to spill every time. I think this will need to be configurable behavior in winch, but it greatly simplifies knowing where locals live.

Reply:
I like this plan and agree with it too; I think there's a distinction or principle that might be worth making explicit in the RFC related to it:

One interesting constraint that the debug support in Cranelift traditionally had was that it must not modify the generated code; in other words, the user should be able to witness a strange behavior, attach a debugger, and be confident they are debugging exactly the same code. That constraint filtered through to how regalloc2 handles debug maps (e.g., disallows a "spill everything between instructions" approach) and consequently led to a *lot* of complexity.

In hindsight I think that was focusing on the wrong thing, because it was implicitly assuming that the sort of bugs the user would be debugging would have to do with Cranelift optimizations interacting with their code. (E.g., if you're gdb'ing a native program, and suspect UB-related issues in your C++, you really do want to debug the exact machine code with the issue.) But what we care about here is the Wasm virtual machine-level behavior and we're free to swap out the implementation.

I bring all this up because I think there might still be folks surprised by an "insert hooks" approach, with as much new plumbing and deoptimization as needed between Wasm instructions, and it seems worthwhile to state what we're preserving (Wasm-level semantics) and why this is OK to do.

Reply:
For cg_clif it is still important for debug info to not cause a huge perf hit as full debug info is enabled by default. Note that cg_clif currently doesn't support emitting debug info for locals though.

Reply:
There are some other tradeoffs we should mention around this choice too, I think. One aspect of the opposite end of the spectrum, the "must not modify machine code" approach, is that it *can* work even on release builds, in a best-effort way (values may be the infamous "`<optimized out>`"). This is sometimes desirable in large programs that take a while to get to the point of interest; if the continues between breakpoints are compute-intensive, basically. (The few times I've gdb'd Cranelift itself I did id on release builds!)

I think the general question to answer is the tradeoff between performance-under-debug and... all the other metrics of goodness (debug info richness, ease of debugger implementation, etc). We've swung very much to the latter here. How do address use-cases that need the former?

One answer could be that record-and-replay addresses one common cause for that need: bug occurs very far into program execution. One could think of it as fast-forwarding with optimized code then switching (by loading the R&R snapshot and entering the debugger) to debug-mode code. To be ideally useful, we'd want R&R support for Cranelift-compiled code too. Is that tentatively part of the plan? (Maybe, if most of the logic is in Wasmtime imports/exports?)

Another answer could be on-stack replacement (OSR), and IIRC this is what SpiderMonkey does when a debug breakpoint is hit. But de-opt'ing from Cranelift to Winch code is an extremely hard problem to think about so maybe let's not cross that bridge yet!

Anyway, my opinions/setpoints on the above are: we should mention that perf is sometimes important; we should name record & replay as the answer when it's needed; we should eventually plan to implement R&R support for Cranelift-compiled code, after we do it for Winch.

Reply:
I wasn't even thinking about deoptimizations between Winch and Cranelift; mostly about Winch so we could have code without the hooks unless we need them.  We'd need to measure the performance impact of having those hooks, though, because it might be OK (and it'll have to be from the get go!).

At a much later date, however, yes, if we're debugging with Cranelift being used as a JIT compiler, it might be worth investigating using Winch as a debugging tier, or even as a tier 0 compiler like other JIT compilers do.

But one thing at a time!

Reply:
To be clear, my suggestion is not to do that. Rather I am suggesting to describe this tradeoff in the RFC, and point to R&R as a solution for the "fast-forward to the bug" subproblem of the general performance problem.

Reply:
I tried to capture this with an additional paragraph after these two bulleted points. What do you think?